sys/mips/mips/pmap.c

```diff
@@ -3517,6 +3638,66 @@ pmap_emulate_modified(pmap_t pmap, vm_offset_t va)
 	return (0);
 }
 
+#ifdef CPU_CHERI
+/*
+ * Are any of this page's PTE_SC (cap-store inhibit) bits clear?
+ *
+ * XREF pmap_is_modified
+ */
+boolean_t
+pmap_is_capdirty(vm_page_t m)
+{
+	boolean_t found;
+
+	if (!pmap_page_is_write_mapped(m))
+		return false;
+
+	if (vm_page_astate_load(m).flags & PGA_CAPDIRTY)
+		return true;
+
+	rw_wlock(&pvh_global_lock);
+	found = pmap_testbit_inv(m, PTE_SC);
+	rw_wunlock(&pvh_global_lock);
+
+	return found;
+}
+
+/* XREF pmap_emulate_modified */
+int
+pmap_emulate_capdirty(pmap_t pmap, vm_offset_t va)
+{
+	pt_entry_t *pte;
+
+	PMAP_LOCK(pmap);
+	pte = pmap_pte(pmap, va);
+	if (pte == NULL)
+		panic("pmap_emulate_capdirty: can't find PTE");
+	if (!pte_test(pte, PTE_MANAGED))
+		panic("pmap_emulate_capdirty: unmanaged page");
+
+#ifdef SMP
+	/* It is possible that some other CPU dirtied it */
+	if (!pte_test(pte, PTE_V) || !pte_test(pte, PTE_SC)) {
+		tlb_update(pmap, va, *pte);
+		PMAP_UNLOCK(pmap);
+		return (0);
+	}
+#else
+	if (!pte_test(pte, PTE_V) || !pte_test(pte, PTE_SC))
+		panic("pmap_emulate_capdirty: invalid pte");
+#endif
+	if (pte_test(pte, PTE_RO) || pte_test(pte, PTE_CRO)) {
+		PMAP_UNLOCK(pmap);
+		return (1);
+	}
+
+	pte_clear(pte, PTE_SC);
```

Comment:
I am sure there's a good reason not to, but for my own education: would it be possible to update flags in the backing page from this context? It would be handy to set Dirty and MCC flags from here and not have to sync later.

Reply:
Also, if we're using PTE_LC as a debugging aid rather than to represent the permission, we should clear it here.

Reply:
To be honest, I'm not quite sure.  I was just following `pmap_emulate_modified`.  My initial guess was that we'd have to hold more locks to get back to the `vm_page_t`, but no, it's just `TLBLO_PTE_TO_PA` and `PHYS_TO_VM_PAGE`... ah, on MIPS, `PHYS_TO_VM_PAGE` is (somewhat) expensive because it's a `VM_PHYSSEG_SPARSE` architecture, so we have to walk an array of segments and find the one containing the physical frame in question.  That motivates the "let these flags accumulate in the emulated page tables" approach well enough, I suppose.